[SPARK-51092][SS] Skip the v1 FlatMapGroupsWithState tests with timeout on big endian platforms

jonathan-albrecht-ibm opened the pull request:
### What changes were proposed in this pull request?
Skip the v1 FlatMapGroupsWithState tests with timeout on big endian platforms.

### Why are the changes needed?
The timestampTimeoutAttribute of StateManagerImplV1 is declared as IntegerType instead of LongType which breaks serialization on big endian platforms. This can't be fixed because it would be a breaking schema change so skip the tests instead.

### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Tested with existing tests on amd64 (little endian) and s390x (big endian)

Below is the test result from s390x:

```
- flatMapGroupsWithState - streaming with processing time timeout - state format version 1 !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:471)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 2
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 1 (RocksDBStateStore) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:471)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 1 (RocksDBStateStore with changelog checkpointing) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:471)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 2 (RocksDBStateStore)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 2 (RocksDBStateStore with changelog checkpointing)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 1 !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:539)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 2
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 1 (RocksDBStateStore) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:539)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 1 (RocksDBStateStore with changelog checkpointing) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:539)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 2 (RocksDBStateStore)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 2 (RocksDBStateStore with changelog checkpointing)
--
- flatMapGroupsWithState, state ver 1 !!! CANCELED !!!
  java.nio.ByteOrder.nativeOrder().equals(java.nio.ByteOrder.LITTLE_ENDIAN) was false (StateDataSourceReadSuite.scala:802)
--
- flatMapGroupsWithState, state ver 2
--
- flatMapGroupsWithState, state ver 1 !!! CANCELED !!!
  java.nio.ByteOrder.nativeOrder().equals(java.nio.ByteOrder.LITTLE_ENDIAN) was false (StateDataSourceReadSuite.scala:802)
--
- flatMapGroupsWithState, state ver 2
--
- flatMapGroupsWithState, state ver 1 !!! CANCELED !!!
  java.nio.ByteOrder.nativeOrder().equals(java.nio.ByteOrder.LITTLE_ENDIAN) was false (StateDataSourceReadSuite.scala:802)
--
- flatMapGroupsWithState, state ver 2
```

### Was this patch authored or co-authored using generative AI tooling?
No

jonathan-albrecht-ibm committed:
Change the `dataType` of the `timestampTimeoutAttribute` from IntegerType to LongType
to match the type of the timestamp.

This fixes the v1 state manager to work on big endian platforms. It fixes all failing
v1 state manager related tests on big endian platforms that do not load checkpoint
state that was generated in little endian format.

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/FlatMapGroupsWithStateExecHelper.scala:
```diff
@@ -152,7 +152,7 @@ object FlatMapGroupsWithStateExecHelper {
 
     private val timestampTimeoutAttribute =
-      AttributeReference("timeoutTimestamp", dataType = IntegerType, nullable = false)()
+      AttributeReference("timeoutTimestamp", dataType = LongType, nullable = false)()
```
@HeartSaVioR Could you take a look at this, please.

HeartSaVioR commented:
Thanks for the contribution!

I'm afraid I have to say we can't fix that. The version is something we leave for backward compatibility. If you start a new streaming query, it will automatically pick the v2. That said, it's not an existing issue but the streaming query with old checkpoint (maybe 2.x era) would have it. And fixing this may lead the existing checkpoint in old era to stop working. 
(I suspect this change to influence state schema, which is a breaking change.)

So let's move on to the latest version and see whether there is a problem.

jonathan-albrecht-ibm commented:
@HeartSaVioR thanks for reviewing. I understand that a schema change could break backwards compatibility. I'd just like to mention in case it helps that all of the tests, except the two in  FlatMapGroupsWithStateExecHelperSuite.scala that explicitly assert on the schema, continue to pass. That includes tests that load flatmapgroup checkpoint files from 2.x era. Would it be possible to write a test or something that would show this change is safe? I'd be happy to work on it with some guidance.

> So let's move on to the latest version and see whether there is a problem.

I'm not sure what `latest version` you mean above but if you mean the v2 version of the state manager then I can say from my testing that the v2 state manager passes all tests on big endian platforms and I don't know of any problems with it.

HeartSaVioR commented:
Since we can't make a change to the schema for the existing version, let's simply mark failing tests (or all tests running with v1) to be ignored with big-endian system. We can put assume() to skip tests; I'm not sure there is a simple way to check the endianness of the system (I haven't had a chance to work across these systems), but there should be a way to do.

jonathan-albrecht-ibm commented:
Ok thanks @HeartSaVioR. I'll update this PR to skip the tests on big endian platforms. I'll try to have something ready next week.

jonathan-albrecht-ibm committed:
Revert "Change the `dataType` of the `timestampTimeoutAttribute` from IntegerType to LongType"

This reverts commit 57487e58116577fb09cc1255ca14100f94f901ab.

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm committed:
Skip the v1 FlatMapGroupsWithState tests with timeout on big endian platforms because the
timestampTimeoutAttribute of StateManagerImplV1 is declared as IntegerType instead
of LongType which breaks serialization on big endian. This can't be fixed because it
would be a breaking schema change.

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm commented:
@HeartSaVioR, I have updated this PR to just skip the failing v1 tests on big endian platforms

HeartSaVioR commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala:
```diff
@@ -794,6 +795,11 @@ abstract class StateDataSourceReadSuite extends StateDataSourceTestBase with Ass
+    // StateManagerImplV1 is declared as IntegerType instead of LongType which breaks
+    // serialization on big endian. This can't be fixed because it would be a breaking
+    // schema change.
+    assume(ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN))
```
nit: `ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)`

maybe good enough to put this to Utils?
cc. @jiangxb1987 @Ngone51 WDYT?

HeartSaVioR commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala:
```diff
@@ -458,7 +459,16 @@ class FlatMapGroupsWithStateSuite extends StateStoreMetricsTest {
+  // of LongType which breaks serialization on big endian. This can't be fixed because it
+  // would be a breaking schema change.
+  def isStateFormatSupported(stateFormatVersion: Int): Boolean = {
+    stateFormatVersion != 1 || ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)
```
ditto

HeartSaVioR reviewed (commented):
Could you please update the PR title and description, as well as JIRA? Especially for PR - we are using the title and description to fill out commit title/message.

Also, looks like linter is failing. Could you please fix it via running `dev/scalastyle` and fixing them? CI also gives you the error.

It'd be awesome if you could post the test run for these test suites to verify they are excluded, because I (and CI) can't verify the fix via running tests on that system.

jonathan-albrecht-ibm committed:
Fix line too long formatting

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm commented:
Thanks for reviewing @HeartSaVioR 
> Could you please update the PR title and description, as well as JIRA? Especially for PR - we are using the title and description to fill out commit title/message.

🤦 I didn't think of this, thx. done
> 
> Also, looks like linter is failing. Could you please fix it via running `dev/scalastyle` and fixing them? CI also gives you the error.

🤦 Thanks for pointing me to dev/scalastyle. done
> 
> It'd be awesome if you could post the test run for these test suites to verify they are excluded, because I (and CI) can't verify the fix via running tests on that system.

Here are the lines from the test run on s390x showing that the v1 tests are excluded and the v2 tests are included:

```txt
- flatMapGroupsWithState - streaming with processing time timeout - state format version 1 !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:471)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 2
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 1 (RocksDBStateStore) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:471)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 1 (RocksDBStateStore with changelog checkpointing) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:471)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 2 (RocksDBStateStore)
--
- flatMapGroupsWithState - streaming with processing time timeout - state format version 2 (RocksDBStateStore with changelog checkpointing)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 1 !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:539)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 2
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 1 (RocksDBStateStore) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:539)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 1 (RocksDBStateStore with changelog checkpointing) !!! CANCELED !!!
  FlatMapGroupsWithStateSuite.this.isStateFormatSupported(FlatMapGroupsWithStateSuite.this.sqlConf.getConf[Int](org.apache.spark.sql.internal.SQLConf.FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION)) was false (FlatMapGroupsWithStateSuite.scala:539)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 2 (RocksDBStateStore)
--
- flatMapGroupsWithState - streaming w/ event time timeout + watermark - state format version 2 (RocksDBStateStore with changelog checkpointing)
--
- flatMapGroupsWithState, state ver 1 !!! CANCELED !!!
  java.nio.ByteOrder.nativeOrder().equals(java.nio.ByteOrder.LITTLE_ENDIAN) was false (StateDataSourceReadSuite.scala:802)
--
- flatMapGroupsWithState, state ver 2
--
- flatMapGroupsWithState, state ver 1 !!! CANCELED !!!
  java.nio.ByteOrder.nativeOrder().equals(java.nio.ByteOrder.LITTLE_ENDIAN) was false (StateDataSourceReadSuite.scala:802)
--
- flatMapGroupsWithState, state ver 2
--
- flatMapGroupsWithState, state ver 1 !!! CANCELED !!!
  java.nio.ByteOrder.nativeOrder().equals(java.nio.ByteOrder.LITTLE_ENDIAN) was false (StateDataSourceReadSuite.scala:802)
--
- flatMapGroupsWithState, state ver 2
```

HeartSaVioR reviewed (approved):
+1

HeartSaVioR commented:
Thanks! Merging to master/4.0.

jonathan-albrecht-ibm commented:
@HeartSaVioR, thanks for reviewing and merging!
